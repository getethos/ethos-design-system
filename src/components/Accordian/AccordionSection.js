import React, { useRef, useEffect } from 'react'
import PropTypes from 'prop-types'
import styles from './AccordionSection.module.css'
import { useAccordionContext } from './useAccordionContext.js'
import { codes } from '../../helpers/constants.js'

export const AccordionSection = ({ children, title, renderToggle, index }) => {
  const {
    focusRef,
    selected,
    expandedAll,
    onToggle,
    onNavigation,
    id,
  } = useAccordionContext()
  const sectionId = `${id}-${index}-section`
  const labelId = `${id}-${index}-label`
  const expanded = expandedAll[index]
  const labelRef = useRef()

  useEffect(() => {
    if (index === selected[0] && labelRef.current) {
      labelRef.current.focus()
    }
  }, [index, selected])

  const onKeyDown = (e) => {
    switch (e.keyCode) {
      case codes.SPACE:
      case codes.RETURN:
        e.preventDefault()
        onToggle && onToggle(index)
        break
      case codes.DOWN:
        e.preventDefault()
        onNavigation(codes.DOWN)
        break
      case codes.UP:
        e.preventDefault()
        onNavigation(codes.UP)
        break
      case codes.HOME:
        e.preventDefault()
        onNavigation(codes.HOME)
        break
      case codes.END:
        e.preventDefault()
        onNavigation(codes.END)
        break
      default:
    }
  }

  const onFocus = () => {
    focusRef.current = index
  }

  const onBlur = () => {
    focusRef.current = null
  }

  const onClick = () => {
    onToggle && onToggle(index)
  }

  return (
    <>
      <div
        role="button"
        aria-expanded={expanded}
        aria-controls={sectionId}
        id={labelId}
        tabIndex={0}
        className={styles.Label}
        onClick={onClick}
        onKeyDown={onKeyDown}
        onFocus={onFocus}
        onBlur={onBlur}
        ref={labelRef}
      >
        {title}
        {renderToggle(expanded)}
      </div>
      <div
        role="region"
        aria-labelledby={labelId}
        id={sectionId}
        hidden={!expanded}
        className={styles.Panel}
      >
        {children}
      </div>
    </>
  )
}

AccordionSection.propTypes = {
  index: PropTypes.number,
  title: PropTypes.string.isRequired,
  renderToggle: PropTypes.func,
  children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
}
