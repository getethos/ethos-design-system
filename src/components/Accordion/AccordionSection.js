import React, { useRef, useEffect } from 'react'
import PropTypes from 'prop-types'
import styles from './AccordionSection.module.css'
import { useAccordionContext } from './useAccordionContext.js'
import { codes } from '../../helpers/constants.js'

/**
 * HeaderBar will place all the event handlers on the header itself,
 * if `toggleChildIsTarget` is false. If `toggleChildIsTarget` is true,
 * we put those handlers ont the <Toggler /> itself.
 */
const HeaderBar = ({
  renderToggle,
  focusRef,
  onToggle,
  toggleChildIsTarget,
  onNavigation,
  expanded,
  sectionId,
  labelId,
  labelRef,
  index,
  title,
}) => {
  const onKeyDown = (e) => {
    switch (e.keyCode) {
      case codes.SPACE:
      case codes.RETURN:
        e.preventDefault()
        onToggle && onToggle(index)
        break
      case codes.DOWN:
        e.preventDefault()
        onNavigation(codes.DOWN)
        break
      case codes.UP:
        e.preventDefault()
        onNavigation(codes.UP)
        break
      case codes.HOME:
        e.preventDefault()
        onNavigation(codes.HOME)
        break
      case codes.END:
        e.preventDefault()
        onNavigation(codes.END)
        break
      default:
    }
  }

  const onFocus = () => {
    focusRef.current = index
  }

  const onBlur = () => {
    focusRef.current = null
  }

  const onClick = () => {
    onToggle && onToggle(index)
  }

  /**
   * It's important here that we've done the following:
   * 1. Have HeaderBar outside of AccordionSection
   * 2. Not have multiple return statements which trigger rerenders
   * https://medium.com/@cowi4030/optimizing-conditional-rendering-in-react-3fee6b197a20
   * All of these allow focus to be preserved as we toggle the AccordianSection
   * since we aren't triggering needless rerenders.
   */
  return (
    <>
      {toggleChildIsTarget && (
        <div className={styles.Label}>
          {title}
          <span
            role="button"
            aria-expanded={expanded}
            aria-controls={sectionId}
            id={labelId}
            tabIndex={0}
            className={styles.Label}
            onClick={onClick}
            onKeyDown={onKeyDown}
            onFocus={onFocus}
            onBlur={onBlur}
            ref={labelRef}
          >
            {renderToggle(expanded)}
          </span>
        </div>
      )}
      {!toggleChildIsTarget && (
        <div
          role="button"
          aria-expanded={expanded}
          aria-controls={sectionId}
          id={labelId}
          tabIndex={0}
          className={styles.Label}
          onClick={onClick}
          onKeyDown={onKeyDown}
          onFocus={onFocus}
          onBlur={onBlur}
          ref={labelRef}
        >
          {title}
          {renderToggle(expanded)}
        </div>
      )}
    </>
  )
}

HeaderBar.propTypes = {
  index: PropTypes.number,
  title: PropTypes.string.isRequired,
  renderToggle: PropTypes.func,
  focusRef: PropTypes.object,
  onToggle: PropTypes.func,
  toggleChildIsTarget: PropTypes.bool,
  onNavigation: PropTypes.func,
  expanded: PropTypes.bool,
  sectionId: PropTypes.string,
  labelId: PropTypes.string,
  labelRef: PropTypes.object,
}

export const AccordionSection = ({ children, title, renderToggle, index }) => {
  const {
    focusRef,
    selected,
    expandedAll,
    onToggle,
    toggleChildIsTarget,
    onNavigation,
    id,
  } = useAccordionContext()
  const sectionId = `${id}-${index}-section`
  const labelId = `${id}-${index}-label`
  const expanded = expandedAll[index]
  const labelRef = useRef()

  useEffect(() => {
    if (index === selected[0] && labelRef.current) {
      labelRef.current.focus()
    }
  }, [index, selected])

  return (
    <>
      <HeaderBar
        renderToggle={renderToggle}
        focusRef={focusRef}
        onToggle={onToggle}
        toggleChildIsTarget={toggleChildIsTarget}
        onNavigation={onNavigation}
        expanded={expanded}
        sectionId={sectionId}
        labelId={labelId}
        labelRef={labelRef}
        index={index}
        title={title}
      />
      <div
        role="region"
        aria-labelledby={labelId}
        id={sectionId}
        hidden={!expanded}
        className={styles.Panel}
      >
        {children}
      </div>
    </>
  )
}

AccordionSection.propTypes = {
  index: PropTypes.number,
  title: PropTypes.string.isRequired,
  renderToggle: PropTypes.func,
  children: PropTypes.node.isRequired,
}
