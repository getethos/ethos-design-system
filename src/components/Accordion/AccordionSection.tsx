import * as React from 'react'
import { useEffect, useRef } from 'react'
import { HeaderBar } from './HeaderBar'
import { useAccordionContext } from './useAccordionContext'
type AccordionSectionProps = {
  index?: number
  title: string
  renderToggle: (expanded: boolean) => JSX.Element
  labelClassName?: string
  panelClassName?: string
  toggleClassName?: string
}
export const AccordionSection: React.FC<AccordionSectionProps> = ({
  children,
  title,
  renderToggle,
  index = 0,
  labelClassName = '',
  panelClassName = '',
  toggleClassName = '',
}) => {
  const {
    focusRef,
    selected,
    expandedAll,
    onToggle,
    toggleChildIsTarget,
    onNavigation,
    id,
  } = useAccordionContext()
  const sectionId = `${id}-${index}-section`
  const labelId = `${id}-${index}-label`
  const expanded = expandedAll[index]
  const labelRef = useRef<any>()
  useEffect(() => {
    if (index === selected[0] && labelRef.current) {
      labelRef?.current?.focus()
    }
  }, [index, selected])
  return (
    <>
      <HeaderBar
        renderToggle={renderToggle}
        focusRef={focusRef}
        onToggle={onToggle}
        toggleChildIsTarget={toggleChildIsTarget}
        onNavigation={onNavigation}
        expanded={expanded}
        sectionId={sectionId}
        labelId={labelId}
        labelRef={labelRef}
        index={index}
        title={title}
        labelClassName={labelClassName}
        toggleClassName={toggleClassName}
      />
      <div
        role="region"
        aria-labelledby={labelId}
        id={sectionId}
        hidden={!expanded}
        className={panelClassName}
      >
        {children}
      </div>
    </>
  )
}
