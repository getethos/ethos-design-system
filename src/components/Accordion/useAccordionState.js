import { useState } from 'react'

export const useAccordionState = (initialState) => {
  const [expanded, setExpanded] = useState(initialState)
  const onToggle = (id) => {
    setExpanded({
      ...expanded,
      [id]: !expanded[id],
    })
  }
  return { expanded, onToggle }
}
