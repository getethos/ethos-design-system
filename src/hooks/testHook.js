/**
* Hooks cannot be called from anywhere like regular javascript
* functions. They can only be called from a functional component
* or from other hooks. This gets around it and the Invariant Violation
* when testing custom hooks that in turn call hook e.g. useState
*/
import React from 'react'
import renderer from 'react-test-renderer'

const TestHook = ({ callback }) => {
  callback()
  return null
}

export default (callback) => {
  renderer.create(<TestHook callback={callback} />)
}
