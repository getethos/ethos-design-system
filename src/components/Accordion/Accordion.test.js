import React from 'react'
import { Accordion } from './index'
import { AccordionSection } from './AccordionSection'
import { act } from 'react-test-renderer'
import testHook from '../../hooks/testHook.js'
import { useAccordionState } from './useAccordionState'
import { render } from '@testing-library/react'

const toggleChild = (expanded) => {
  return <span aria-hidden={true}>{expanded ? '▲' : '▼'}</span>
}

describe('Accordion', () => {
  it('default rendering', async () => {
    await act(async () => {
      testHook(() => {
        const { expanded, onToggle } = useAccordionState({ 1: true })
        const tree = render(
          <Accordion expanded={expanded} onToggle={onToggle}>
            <AccordionSection
              title="Section 1"
              id="id1"
              renderToggle={toggleChild}
            >
              <p>para</p>
              <span>span</span>
            </AccordionSection>
            <AccordionSection
              title="Section 2"
              id="id2"
              renderToggle={toggleChild}
            >
              <p>para</p>
              <span>span</span>
            </AccordionSection>
          </Accordion>
        )
        expect(tree).toMatchSnapshot()
      })
    })
  })
})
