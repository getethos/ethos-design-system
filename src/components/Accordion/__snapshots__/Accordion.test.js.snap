// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordian default rendering 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          aria-controls="undefined-0-section"
          class="Label"
          id="undefined-0-label"
          role="button"
          tabindex="0"
        >
          Section 1
          <span
            aria-hidden="true"
          >
            ▼
          </span>
        </div>
        <div
          aria-labelledby="undefined-0-label"
          class="Panel"
          hidden=""
          id="undefined-0-section"
          role="region"
        >
          <p>
            para
          </p>
          <span>
            span
          </span>
        </div>
        <div
          aria-controls="undefined-1-section"
          aria-expanded="true"
          class="Label"
          id="undefined-1-label"
          role="button"
          tabindex="0"
        >
          Section 2
          <span
            aria-hidden="true"
          >
            ▲
          </span>
        </div>
        <div
          aria-labelledby="undefined-1-label"
          class="Panel"
          id="undefined-1-section"
          role="region"
        >
          <p>
            para
          </p>
          <span>
            span
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        aria-controls="undefined-0-section"
        class="Label"
        id="undefined-0-label"
        role="button"
        tabindex="0"
      >
        Section 1
        <span
          aria-hidden="true"
        >
          ▼
        </span>
      </div>
      <div
        aria-labelledby="undefined-0-label"
        class="Panel"
        hidden=""
        id="undefined-0-section"
        role="region"
      >
        <p>
          para
        </p>
        <span>
          span
        </span>
      </div>
      <div
        aria-controls="undefined-1-section"
        aria-expanded="true"
        class="Label"
        id="undefined-1-label"
        role="button"
        tabindex="0"
      >
        Section 2
        <span
          aria-hidden="true"
        >
          ▲
        </span>
      </div>
      <div
        aria-labelledby="undefined-1-label"
        class="Panel"
        id="undefined-1-section"
        role="region"
      >
        <p>
          para
        </p>
        <span>
          span
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
